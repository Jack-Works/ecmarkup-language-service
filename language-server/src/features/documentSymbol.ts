import { isElementNode } from 'parse5/lib/tree-adapters/default.js'
import {
    type CancellationToken,
    type Connection,
    type DocumentSymbol,
    type DocumentSymbolClientCapabilities,
    type DocumentSymbolParams,
    Location,
    type ResultProgressReporter,
    type ServerCapabilities,
    type SymbolInformation,
    SymbolKind,
    type WorkDoneProgressReporter,
} from 'vscode-languageserver'
import type { TextDocument } from 'vscode-languageserver-textdocument'
import { getAbstractOperationHeader } from '../parser/ecmarkup.js'
import { documents } from '../server-shared.js'
import { formatSpecValueText } from '../utils/format.js'
import type { Program } from '../workspace/program.js'

export class DocumentSymbols {
    constructor(
        public capabilities: DocumentSymbolClientCapabilities | undefined = { hierarchicalDocumentSymbolSupport: true },
    ) {}

    async findDocumentSymbols(
        document: TextDocument,
        program: Program,
        _params: DocumentSymbolParams,
        _token?: CancellationToken | undefined,
        _workDoneProgress?: WorkDoneProgressReporter | undefined,
        _resultProgress?: ResultProgressReporter<SymbolInformation[] | DocumentSymbol[]> | undefined,
    ): Promise<SymbolInformation[] | DocumentSymbol[] | null | undefined> {
        const sourceFile = program.getSourceFile(document)
        const symbols: DocumentSymbol[] = []
        for (const grammar of sourceFile.localDefinedGrammars) {
            symbols.push({
                name: grammar.name,
                kind: SymbolKind.Interface,
                range: sourceFile.getRelativeRange(grammar.node, grammar.fullDefinitionRange),
                selectionRange: sourceFile.getRelativeRange(grammar.node, grammar.range),
            })
        }
        for (const operation of sourceFile.localDefinedAbstractOperations) {
            const header = getAbstractOperationHeader(operation.node)
            const position = operation.range.position + operation.range.length
            const headerText = sourceFile.getNodeInnerText(header)?.slice(position)
            let detail: string | undefined
            const seenVariables = new Set<string>()
            const children: DocumentSymbol[] = []

            if (header && headerText) {
                const params = [...headerText.matchAll(/_(\w+)_/g)]
                params.forEach((param) => {
                    if (!seenVariables.has(param[1]!)) {
                        seenVariables.add(param[1]!)
                        const range = sourceFile.getRelativeRangeToInnerText(
                            header,
                            position + param.index + 1,
                            param[1]!.length,
                        )
                        children.push({
                            name: param[1]!,
                            kind: SymbolKind.Variable,
                            range,
                            selectionRange: range,
                        })
                    }
                })
                detail = `(${params.map((param) => param[1]).join(', ')})`
                const returnType = headerText.split(':').at(-1)
                if (returnType) detail += `: ${formatSpecValueText(returnType.trim()).trim()}`
            }

            const parent = operation.node.parentNode
            const source = sourceFile.getNodeInnerText(parent)
            if (isElementNode(parent) && source) {
                for (const match of source.matchAll(/let\s*_(\w+)_/dgi)) {
                    if (!seenVariables.has(match[1]!)) {
                        seenVariables.add(match[1]!)
                        const range = sourceFile.getRelativeRangeToInnerText(
                            parent,
                            match.indices![1]![0],
                            match[1]!.length,
                        )
                        children.push({
                            name: match[1]!,
                            kind: SymbolKind.Variable,
                            range,
                            selectionRange: range,
                        })
                    }
                }
            }

            symbols.push({
                name: operation.name,
                kind: SymbolKind.Function,
                range: sourceFile.getRelativeRange(operation.node, operation.fullDefinitionRange),
                selectionRange: sourceFile.getRelativeRange(operation.node, operation.range),
                detail,
                children,
            })
        }
        return this.toClient(document.uri, symbols)
    }

    private toClient(uri: string, result: SymbolInformation[] | DocumentSymbol[]) {
        if (this.capabilities?.hierarchicalDocumentSymbolSupport) {
            return result
        } else {
            return (result as DocumentSymbol[]).flatMap((symbol): SymbolInformation => {
                return {
                    kind: symbol.kind,
                    location: Location.create(uri, symbol.selectionRange),
                    name: symbol.name,
                }
            })
        }
    }

    static enable(
        serverCapabilities: ServerCapabilities,
        connection: Connection,
        program: Program,
        capabilities: DocumentSymbolClientCapabilities | undefined,
    ) {
        const provider = new DocumentSymbols(capabilities)
        connection.onDocumentSymbol((params, token, workDoneProgress, resultProgress) => {
            const document = documents.get(params.textDocument.uri)
            if (!document) return undefined
            return provider.findDocumentSymbols(document, program, params, token, workDoneProgress, resultProgress)
        })
        serverCapabilities.documentSymbolProvider = {}
    }
}
