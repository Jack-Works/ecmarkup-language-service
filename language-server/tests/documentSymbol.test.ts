import { expect, it } from 'vitest'
import type { TextDocument } from 'vscode-languageserver-textdocument'
import { type DocumentSymbol, type SymbolInformation, SymbolKind } from 'vscode-languageserver-types'
import { DocumentSymbolProvider } from '../src/features/documentSymbol.js'
import { betterSnapshot, File, type MarkedOffset, type MarkedRange } from './File.js'

it('provides document symbols', async () => {
    const documentSymbol = new DocumentSymbolProvider()
    const { document, textDocument, program } = File.basic`
        <!doctype html>
        <emu-clause id="sec-OpenCalc" type="abstract operation">
            <h1>
                OpenCalc (
                    _input_: an ECMAScript language value,
                    _input2_: an ECMAScript language value,
                    [ _input3_: an ECMAScript language value, ]
                ): either a normal completion containing an Object or a throw completion
            </h1>
            <dl class="header">
            </dl>
            <emu-alg>
                1. Let _x_ be ? ToObject(_input_).
                2. Let _y_ be ? ToObject(_input2_).
                3. Set _y_ to ? ToObject(_input2_).
                4. Return the result of calling User Code with _x_ as this value and _y_ as the first argument.
            </emu-alg>
        </emu-clause>
    `
    const result = await documentSymbol.findDocumentSymbols(document, program, { textDocument })
    expect(printSymbols(document, result!)).toMatchInlineSnapshot(`
      "<!doctype html>
      <emu-clause id="sec-OpenCalc" type="abstract operation">
          <h1>
              OpenCalc (
              ~~~~~~~~ OpenCalc (Function) (input, input2, input3): NormalCompletion<an Object> | ThrowCompletion
                  _input_: an ECMAScript language value,
                   ~~~~~ input (Variable)
                  _input2_: an ECMAScript language value,
                   ~~~~~~ input2 (Variable)
                  [ _input3_: an ECMAScript language value, ]
                     ~~~~~~ input3 (Variable)
              ): either a normal completion containing an Object or a throw completion
          </h1>
          <dl class="header">
          </dl>
          <emu-alg>
              1. Let _x_ be ? ToObject(_input_).
                      ~ x (Variable)
              2. Let _y_ be ? ToObject(_input2_).
                      ~ y (Variable)
              3. Set _y_ to ? ToObject(_input2_).
              4. Return the result of calling User Code with _x_ as this value and _y_ as the first argument.
          </emu-alg>
      </emu-clause>"
    `)
    expect(result).toMatchSnapshot()

    documentSymbol.capabilities = undefined
    const result2 = await documentSymbol.findDocumentSymbols(document, program, { textDocument })
    expect(printSymbols(document, result2!)).toMatchInlineSnapshot(`
      "<!doctype html>
      <emu-clause id="sec-OpenCalc" type="abstract operation">
          <h1>
              OpenCalc (
              ~~~~~~~~ OpenCalc (Function)
                  _input_: an ECMAScript language value,
                  _input2_: an ECMAScript language value,
                  [ _input3_: an ECMAScript language value, ]
              ): either a normal completion containing an Object or a throw completion
          </h1>
          <dl class="header">
          </dl>
          <emu-alg>
              1. Let _x_ be ? ToObject(_input_).
              2. Let _y_ be ? ToObject(_input2_).
              3. Set _y_ to ? ToObject(_input2_).
              4. Return the result of calling User Code with _x_ as this value and _y_ as the first argument.
          </emu-alg>
      </emu-clause>"
    `)
    expect(result2).toMatchSnapshot()
})

function printSymbols(document: TextDocument, tokens: (DocumentSymbol | SymbolInformation)[]) {
    return betterSnapshot(
        document,
        tokens.flatMap(function toMarked(symbol): (MarkedRange | MarkedOffset)[] {
            if ('location' in symbol) {
                const { location, kind, name } = symbol
                return [
                    {
                        range: location.range,
                        annotate: ` ${name} (${printKind(kind)})`,
                    },
                ]
            } else {
                const { kind, name, selectionRange, detail, children } = symbol
                return [
                    {
                        annotate: ` ${name} (${printKind(kind)}) ${detail || ''}`.trimEnd(),
                        range: selectionRange,
                    } as MarkedRange | MarkedOffset,
                ].concat(children ? children.flatMap(toMarked) : [])
            }
        }),
    )
}

function printKind(symbol: SymbolKind) {
    return Object.entries(SymbolKind).find(([, value]) => value === symbol)?.[0]
}
