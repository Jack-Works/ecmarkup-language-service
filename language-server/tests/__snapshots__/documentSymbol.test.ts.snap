// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`provides document symbols 2`] = `
[
  {
    "children": [
      {
        "kind": 13,
        "name": "input",
        "range": {
          "end": {
            "character": 26,
            "line": 5,
          },
          "start": {
            "character": 21,
            "line": 5,
          },
        },
        "selectionRange": {
          "end": {
            "character": 26,
            "line": 5,
          },
          "start": {
            "character": 21,
            "line": 5,
          },
        },
      },
      {
        "kind": 13,
        "name": "input2",
        "range": {
          "end": {
            "character": 27,
            "line": 6,
          },
          "start": {
            "character": 21,
            "line": 6,
          },
        },
        "selectionRange": {
          "end": {
            "character": 27,
            "line": 6,
          },
          "start": {
            "character": 21,
            "line": 6,
          },
        },
      },
      {
        "kind": 13,
        "name": "input3",
        "range": {
          "end": {
            "character": 29,
            "line": 7,
          },
          "start": {
            "character": 23,
            "line": 7,
          },
        },
        "selectionRange": {
          "end": {
            "character": 29,
            "line": 7,
          },
          "start": {
            "character": 23,
            "line": 7,
          },
        },
      },
      {
        "kind": 13,
        "name": "x",
        "range": {
          "end": {
            "character": 25,
            "line": 13,
          },
          "start": {
            "character": 24,
            "line": 13,
          },
        },
        "selectionRange": {
          "end": {
            "character": 25,
            "line": 13,
          },
          "start": {
            "character": 24,
            "line": 13,
          },
        },
      },
      {
        "kind": 13,
        "name": "y",
        "range": {
          "end": {
            "character": 25,
            "line": 14,
          },
          "start": {
            "character": 24,
            "line": 14,
          },
        },
        "selectionRange": {
          "end": {
            "character": 25,
            "line": 14,
          },
          "start": {
            "character": 24,
            "line": 14,
          },
        },
      },
    ],
    "detail": "(input, input2, input3): NormalCompletion<an Object> | ThrowCompletion",
    "kind": 12,
    "name": "OpenCalc",
    "range": {
      "end": {
        "character": 88,
        "line": 8,
      },
      "start": {
        "character": 16,
        "line": 4,
      },
    },
    "selectionRange": {
      "end": {
        "character": 24,
        "line": 4,
      },
      "start": {
        "character": 16,
        "line": 4,
      },
    },
  },
]
`;

exports[`provides document symbols 4`] = `
[
  {
    "kind": 12,
    "location": {
      "range": {
        "end": {
          "character": 24,
          "line": 4,
        },
        "start": {
          "character": 16,
          "line": 4,
        },
      },
      "uri": "test://1fe3ac1a5f7405c67b56a061ae3d0ebe.emu",
    },
    "name": "OpenCalc",
  },
]
`;
